aten/src/ATen/native/cpu/TensorCompareKernel.cpp

```diff
@@ -301,6 +301,34 @@ static void mode_kernel_impl(
       });
 }
 
+// Default imperative implementation of isin(). Used when the number of test elements is small.
+static void isin_default_kernel_cpu(
+    const Tensor& elements,
+    const Tensor& test_elements,
+    bool invert,
+    const Tensor& out) {
+  // Since test elements is not an input of the TensorIterator, type promotion
+  // must be done manually.
+  ScalarType common_type = promoteTypes(elements.scalar_type(), test_elements.scalar_type());
```

Comment:
Our type promotion rules are pretty complicated with special handling for scalars and 0D tensors.

If this wants to support type promotion then I think it needs to use torch.result_type here to handle the 0D tensor case?

Alternatively it can require that elements and test_elements have the same dtype, but then the scalar wrappers either need different implementations or they need to wrap the scalar properly

ping me if you have more questions about this

Reply:
I do have some questions about this stuff in general. Probably easier to just VC :)

Reply:
Cool I think I understand. There are special cases for scalars vs. 0D tensors, and using `result_type()` instead of `promoteTypes()` can do this properly since it receives the full tensors instead of just their types.

Switching to `result_type()`